dodiscover/constraint/intervention.py

```diff
@@ -0,0 +1,362 @@
+        self.n_ci_tests += self.skeleton_learner_.n_ci_tests
+        return skel_graph, sep_set
+
+    def fit(self, data: List[pd.DataFrame], context: Context):
+        """Learn the relevant causal graph equivalence class.
+
+        From the pairs of datasets, we take all combinations and
+        construct F-nodes corresponding to those.
+
+        Parameters
+        ----------
+        data : List[pd.DataFrame]
+            The list of different datasets assigned to different
+            environments. We assume the first dataset is always
+            observational.
+        context : Context
+            The context with interventional assumptions.
+
+        Returns
+        -------
+        self : PsiFCI
+            The fitted learner.
+        """
+        if not isinstance(data, list):
+            raise RuntimeError("The input datasets must be in a Python list.")
+
+        n_datasets = len(data)
+        n_distributions = context.num_distributions
+
+        if n_datasets != n_distributions:
+            raise RuntimeError(
+                f"There are {n_datasets} passed in, but {n_distributions} "
+                f"total assumed distributions. There must be a matching number of datasets and "
+                f"'context.num_distributions'."
+            )
+
+        return super().fit(data, context)
+
+    def _apply_rule11(self, graph: EquivalenceClass, context: Context) -> Tuple[bool, List]:
+        """Apply "Rule 8" in I-FCI algorithm, which we call Rule 11.
+
+        This orients all edges out of F-nodes. So patterns of the form
+
+        ``('F', 0) *-* 'x'`` will become ``('F', 0) -> 'x'``.
+
+        For original details of the rule, see :footcite:`Kocaoglu2019characterization`.
+
+        Parameters
+        ----------
+        graph : EquivalenceClass
+            The causal graph to apply rules to.
+        context : Context
+            The causal context.
+
+        Returns
+        -------
+        added_arrows : bool
+            Whether or not arrows were added.
+        oriented_edges : List
+            A list of oriented edges.
+
+        References
+        ----------
+        .. footbibliography::
+        """
+        augmented_nodes = context.get_augmented_nodes()
+
+        oriented_edges = []
+        added_arrows = True
+        for node in augmented_nodes:
+            for nbr in graph.neighbors(node):
+                if nbr in augmented_nodes:
+                    continue
+
+                # remove all edges between node and nbr and orient this out
+                graph.remove_edge(node, nbr)
+                graph.remove_edge(nbr, node)
+                graph.add_edge(node, nbr, graph.directed_edge_name)
+                oriented_edges.append((node, nbr))
+        return added_arrows, oriented_edges
+
+    def _apply_rule12(
+        self, graph: EquivalenceClass, u: Column, a: Column, c: Column, context: Context
+    ) -> bool:
+        """Apply "Rule 9" of the I-FCI algorithm.
```

Comment:
Similar to rule 9, say that it is called rule 12 here. Otherwise, it is a bit confusing.

Reply:
Addressed.